### PR TITLE
Universal channel - 1

### PR DESCRIPTION
**Pull request by alfredo-stonk**
Universal channel and IBC middleware implementation.

### What's done

- Universal Channel (UC) handshake
- Support full IBC packet semantics over UC
- Add IBC middleware (MW) contracts/interfaces which allows for composing arbitary MW stack
- Remove IBC fee and Escrow from Dispatcher
- Tests covering features above with `forge test`

Minor changes:
- using `forge test` Github Action, retiring `make test` which uses old Hardhat project setup
- dedup IBC events with IbcEventsEmitter interface
- add modifier functions for MW and IbcDispatcher callback access control
- add VirtualChain testing contracts which simulates separate deployment of contract instances on multiple virtual chains

Docs on Universal Channel and IBC Middleware
`docs/universal-channel-middleware.md` 

### How to test

```sh
forge test
```

Main test file is `test/universal.channel.t.sol`

Prerequisite:
Install Foundry at https://book.getfoundry.sh/getting-started/installation

### TODOs

- e2e test at polymerase repo
- 
Lower priorities
- add ics-29 packet fee MW 
- minor: migrate old Dispatcher tests to using VirtualChain for more test cases

### CODE REVIEW

File: contracts/Dispatcher.sol [hunk 1]
```diff
@@ -494,39 +430,86 @@ contract Dispatcher is IbcDispatcher, Ownable {
...
+        sendPacketCommitment[sender][channelId][sequence] = true;
         // increment nextSendPacketSequence
-        nextSequenceSend[msg.sender][channelId] = sequence + 1;
+        nextSequenceSend[sender][channelId] = sequence + 1;
+
+        emit SendPacket(sender, channelId, packet, sequence, timeoutTimestamp, fee);
+    }
+
+    /**
+     * @notice Sends an IBC packet over a universal channel with the specified packet data and timeout block timestamp.
```
[the remark this answers is not in the document]
Comment (alfredo-stonk) [reply contracts/Dispatcher.sol]: doc added at `docs/universal-channel-middleware.md`

File: contracts/UniversalChannelHandler.sol [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+contract UniversalChannelHandler is IbcReceiverBase, IbcUniversalChannelMW {
+    constructor(IbcDispatcher _dispatcher) IbcReceiverBase(_dispatcher) {}
+
+    bytes32[] public connectedChannels;
+    string constant VERSION = "1.0";
+    uint256 public constant MW_ID = 1;
+
+    // Key: middleware bitmap, Value: middleware address from receiver(chain B)'s perspective
+    mapping(uint256 => address[]) public mwStackAddrs;
```
Comment (notbdu) [inline contracts/UniversalChannelHandler.sol:19]: Was thinking we'd just pass this in from Polymer?

Also, we'd have to register every combination like this.

File: contracts/UniversalChannelHandler.sol [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        ChannelOrder ordering,
+        bool feeEnabled,
+        string[] calldata connectionHops,
+        string calldata counterpartyPortId,
+        bytes32 counterpartyChannelId,
+        string calldata counterpartyVersion
+    ) external onlyIbcDispatcher returns (string memory selectedVersion) {
+        if (counterpartyChannelId == bytes32(0)) {
+            // ChanOpenInit
+            require(keccak256(abi.encodePacked(version)) == keccak256(abi.encodePacked(VERSION)), "Unsupported version");
```
Comment (notbdu) [inline contracts/UniversalChannelHandler.sol:43]: Is this cheaper than comparing two strings?

File: contracts/IbcMiddleware.sol [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+ */
+interface IbcMiddleware is IbcMiddlwareProvider, IbcMwPacketReceiver, IbcUniversalPacketReceiver {}
+
+/**
+ * @title IbcUniversalChannelMW
+ * @notice This interface must be implemented by IBC universal channel middleware contracts.
+ * IbcUniversalChannelMW is a special type of IbcMiddleware that owns IBC channels, which are multiplexed for other IbcMiddleware.
+ * IbcUniversalChannelMW cannot sit on top of other MW, and must talk to IbcDispatcher directly.
+ */
+interface IbcUniversalChannelMW is IbcMiddlwareProvider, IbcPacketReceiver, IbcChannelReceiver {}
```
Comment (notbdu) [inline contracts/IbcMiddleware.sol:112]: Does this need to be a separate interface if only one contract impls it?

I guess for later versions?

File: contracts/UniversalChannelHandler.sol [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            if (connectedChannels[i] == channelId) {
+                delete connectedChannels[i];
+                channelFound = true;
+                break;
+            }
+        }
+        require(channelFound, "Channel not found");
+    }
+
+    function sendUniversalPacket(
```
Comment (notbdu) [inline contracts/UniversalChannelHandler.sol:81]: Same comment as earlier on combining the send Universal/MW packet methods + rename

File: contracts/Ibc.sol [hunk 6]
```diff
@@ -74,48 +91,79 @@ struct IbcEndpoint {
...
-     * @dev Return the amount of fee to be refunded to the sender when a packet is acknowledged.
-     * @param fee accumulative packet fees
-     */
-    function ackRefundAmount(PacketFee memory fee) internal pure returns (uint256) {
-        if (fee.timeoutFee > fee.ackFee + fee.recvFee) {
-            return fee.timeoutFee - fee.ackFee - fee.recvFee;
-        } else {
-            return 0;
+    // convert an address to its hex string, but without 0x prefix
+    function toHexStr(address addr) internal pure returns (bytes memory) {
```
Comment (notbdu) [inline contracts/Ibc.sol:151]: Is this only used for testing? These ops look expensive.

File: contracts/UniversalChannelHandler.sol [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+    /**
+     * @dev Register a middleware stack for universal packet routing.
+     * This is a temporary solution for testing only.
+     * Polymer chain will maintain a global registry of middleware stacks.
+     * @param mwBitmap Bit OR of all MW IDs in the stack, excluding this MW's ID
+     * @param mwAddrs addresses in the stack, from the perspective of the receiver (chain B)
+     * MW closer to UniversalChannel MW has smaller index. MW stack must be in the same order on both chains.
+     */
+    function registerMwStack(uint256 mwBitmap, address[] calldata mwAddrs) external onlyOwner {
```
Comment (notbdu) [inline contracts/UniversalChannelHandler.sol:167]: Same comment as earlier on passing MW bitmap in versus registering on chain.

File: docs/universal-channel-middleware.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
+# Universal Channel and IBC Middleware
```
Comment (notbdu) [inline docs/universal-channel-middleware.md:1]: Nice docs! 🙌

**Comment by dshiell** [comment]
approving to sync with current deployment. should create follow up issues to address outstanding PR comments.

**Review by dshiell — APPROVED** [review]

**Pull request merged**